webgl: Refactor texture validations to take advantage of rust type system

emilio opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because refactoring.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

This commit introduces the `WebGLValidator` trait, and uses it for multiple
validations in the texture-related WebGL code, to move that logic out of the
already bloated `webglrenderingcontext.rs` file.

It also creates a type-safe wrapper for some WebGL types, removing all the
`unreachable!`s there, and introduces a macro for generating them conveniently.

This partially addresses #10693, pending refactor more code to use this
infrastructure, and (possibly?) introducing an `AsGLError` trait for the errors
to make the error handling happen in `WebGLContext`.

r? @jdm

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11724)

<!-- Reviewable:end -->

emilio committed:
webgl: Refactor texture validations to take advantage of rust type system

This commit introduces the `WebGLValidator` trait, and uses it for multiple
validations in the texture-related WebGL code, to move that logic out of the
already bloated `webglrenderingcontext.rs` file.

It also creates a type-safe wrapper for some WebGL types, removing all the
`unreachable!`s there, and introduces a macro for generating them conveniently.

This partially addresses #10693, pending refactor more code to use this
infrastructure, and (possibly?) introducing an `AsGLError` trait for the errors
to make the error handling happen in `WebGLContext`.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/mod.rs, components/script/dom/webgl_validations/types.rs, components/script/dom/webgl_validations/mod.rs, components/script/dom/webgltexture.rs, components/script/dom/webglrenderingcontext.rs, components/script/dom/webgl_validations/tex_image_2d.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 2d97df6 with merge b148b4e...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1624)

highfive commented:
▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/copyTexImage2DBadArgs.html:
      │ FAIL [expected PASS] WebGL test #0: testTexImage2D
      │   → assert_true: testTexImage2D expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:7
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:899:5
      │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:188:5
      │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:922:5
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:947:7
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/texSubImage2D.html:
      └ PASS [expected FAIL] WebGL test #0: testTexSubImage2D
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/texture-npot.html:
      │ FAIL [expected PASS] WebGL test #1: getError expected: INVALID_VALUE. Was NO_ERROR : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE
      │   → assert_true: getError expected: INVALID_VALUE. Was NO_ERROR : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE expected true got false
      │ FAIL [expected PASS] WebGL test #13: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE
      │   → assert_true: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE expected true got false
      │ FAIL [expected PASS] WebGL test #25: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE
      │   → assert_true: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE expected true got false
      │ FAIL [expected PASS] WebGL test #37: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE
      │   → assert_true: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE expected true got false
      │ FAIL [expected PASS] WebGL test #49: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE
      │   → assert_true: getError expected: INVALID_VALUE. Was INVALID_OPERATION : gl.texImage2D with NPOT texture with level &gt; 0 should return INVALID_VALUE expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
      │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
      │ glErrorShouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1379:5
      │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:67:3
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:60:1
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/texture-npot.html:
      │ FAIL [expected PASS] WebGL test #35: at (0, 0) expected: 192,192,192,255 was 0,0,0,255
      │   → assert_true: at (0, 0) expected: 192,192,192,255 was 0,0,0,255 expected true got false
      │ </span><span class="stdout">
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
      │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
      │ checkCanvasRectColor@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1046:9
      │ checkCanvasRect@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1072:3
      │ checkCanvas@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1095:3
      │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:125:3
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:60:1
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/texture-npot.html:
      │ FAIL [expected PASS] WebGL test #34: getError expected: NO_ERROR. Was INVALID_OPERATION : gl.texImage2D and gl.generateMipmap with POT texture at level 0 should succeed
      │   → assert_true: getError expected: NO_ERROR. Was INVALID_OPERATION : gl.texImage2D and gl.generateMipmap with POT texture at level 0 should succeed expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
      │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
      │ glErrorShouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1379:5
      │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:116:3
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:60:1

highfive commented:
New code was committed to pull request.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit bbc3008 with merge 169d743...

highfive commented:
New code was committed to pull request.

emilio commented:
@bors-servo: try

(typo, sorry for the spam)

bors-servo commented:
:hourglass: Trying commit 5c85052 with merge 056e2ff...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1649)

highfive commented:
▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/copyTexImage2DBadArgs.html:
      │ FAIL [expected PASS] WebGL test #0: testTexImage2D
      │   → assert_true: testTexImage2D expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:900:7
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:899:5
      │ runTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:188:5
      │ initTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:922:5
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/more/unit.js:947:7
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/more/functions/texSubImage2D.html:
      └ PASS [expected FAIL] WebGL test #0: testTexSubImage2D
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/texture-npot.html:
      │ FAIL [expected PASS] WebGL test #34: getError expected: NO_ERROR. Was INVALID_OPERATION : gl.texImage2D and gl.generateMipmap with POT texture at level 0 should succeed
      │   → assert_true: getError expected: NO_ERROR. Was INVALID_OPERATION : gl.texImage2D and gl.generateMipmap with POT texture at level 0 should succeed expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
      │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
      │ glErrorShouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1379:5
      │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:116:3
      └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:60:1
    
      ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/texture-npot.html:
      │ FAIL [expected PASS] WebGL test #35: at (0, 0) expected: 192,192,192,255 was 0,0,0,255
      │   → assert_true: at (0, 0) expected: 192,192,192,255 was 0,0,0,255 expected true got false
      │ 
      │ reportTestResultsToHarness/&lt;@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
      │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
      │ test@http://web-platform.test:8000/resources/testharness.js:495:9
      │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:7
      │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
      │ checkCanvasRectColor@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1046:9
      │ checkCanvasRect@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1072:3
      │ checkCanvas@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1095:3
      │ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/texture-npot.html:125:3
      └ @http://web-platform.test:8000/webgl/conforman</span><span class="stdout">ce-1.0.3/conformance/textures/texture-npot.html:60:1

highfive commented:
New code was committed to pull request.

emilio commented:
Fixed the failures (they were legit ones due to `components()` returning 2 instead of 1 for `Luminance`, that was a typo, and for a general validation that was being done only on `texImage2D`), and changed the expectation for the new passing test (that is also legit \o/).

I *really* need to take a look to some of the other failures we're having, but I'd prefer to check this in for now before digging into it more.

Should be ready for re-review: r? @jdm 

@bors-servo: try

bors-servo commented:
:hourglass: Trying commit f5df6d4 with merge bb8f934...

emilio commented:
BTW, the plan after this is to fill followups for:

  * Moving more things to validators.
  * Give more meaningful errors via `info!` or `warn!` using `error.description()`, and probably merge error handling with `context.webgl_error()` to make it more straightforward and encourage better error descriptions.

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1464), [arm32](http://build.servo.org/builders/arm32/builds/1573), [arm64](http://build.servo.org/builders/arm64/builds/1561), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1595), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1674), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1638), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1622), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1686), [windows](http://build.servo.org/builders/windows/builds/1296)

jdm commented:
This is really great! I've only got a few small comments.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1.
Review status: 4 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/webgltexture.rs, line 307 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL26uzTnrVcJjaJrjB2:-KL26uzTnrVcJjaJrjB3:418137372) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgltexture.rs#L307)):*
> ```Rust
>     }
> 
>     #[inline]
> ```

I'm inclined to leave the inline annotations out unless we come across cases where they actually make a difference.

---

*[components/script/dom/webgl_validations/mod.rs, line 7 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL27DQN2f2LeJWpWIuB:-KL27DQN2f2LeJWpWIuC:-1126308819) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/mod.rs#L7)):*
> ```Rust
> pub trait WebGLValidator {
>     type ValidatedOutput;
>     type Error: ::std::error::Error;
> ```

Do we have plans to do anything with the errors? Right now I think every caller ignores the error value.

---

*[components/script/dom/webgl_validations/tex_image_2d.rs, line 81 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL27UoRnVM_UFCv0t2b:-KL27UoSaZPgaUPREw_G:1816915868) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/tex_image_2d.rs#L81)):*
> ```Rust
> }
> 
> #[inline]
> ```

This seems unnecessary.

---

*[components/script/dom/webgl_validations/tex_image_2d.rs, line 182 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL27zt9meRl1u6tFS6-:-KL27zt9meRl1u6tFS60:464866396) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/tex_image_2d.rs#L182)):*
> ```Rust
>         // // GL_INVALID_VALUE is generated if level is greater than zero and the
>         // // texture is not power of two.
>         // if level > 0 &&
> ```

Why is this commented out?

---

*[components/script/dom/webgl_validations/tex_image_2d.rs, line 336 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL28Q27ttH4XWB8b8g9:-KL28Q28R9-Zh5MBYQK5:-653986567) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/tex_image_2d.rs#L336)):*
> ```Rust
>         }
> ```

nit: remove the excess newlines

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11724#-:-KL28r41CG4sQG9fFX35:1769174625)*
<!-- Sent from Reviewable.io -->

jdm commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11724)*
<!-- Sent from Reviewable.io -->

emilio commented:
Comments adressed :)

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/webgltexture.rs, line 307 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL26uzTnrVcJjaJrjB2:-KL3PeEeSZCZoFYoCfYn:-572314275) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgltexture.rs#L307)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> I'm inclined to leave the inline annotations out unless we come across cases where they actually make a difference.
</details>
Fair enough, done

---

*[components/script/dom/webgl_validations/mod.rs, line 7 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL27DQN2f2LeJWpWIuB:-KL3PfzrUOiEnr1wppto:128000895) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/mod.rs#L7)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Do we have plans to do anything with the errors? Right now I think every caller ignores the error value.
</details>
Yes, ideally we should stop ignoring the errors and stop calling `context.webgl_error(..)` in the validator. Instead, the context would take care of converting `error -> WebGLError` on demand, reporting a useful message given by `error.description()`.

---

*[components/script/dom/webgl_validations/tex_image_2d.rs, line 81 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL27UoRnVM_UFCv0t2b:-KL3Q7TArXOsHxOy7Sho:-499146009) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/tex_image_2d.rs#L81)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> This seems unnecessary.
</details>
Done.

---

*[components/script/dom/webgl_validations/tex_image_2d.rs, line 336 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11724#-KL28Q27ttH4XWB8b8g9:-KL3Q8DaLgHnvDPO3uhJ:-499146009) ([raw file](https://github.com/servo/servo/blob/2d97df6468eacd041f1cb2e1773c0936b76765af/components/script/dom/webgl_validations/tex_image_2d.rs#L336)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> nit: remove the excess newlines
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11724#-:-KL3QbJT5WHhXMG743j4:390071827)*
<!-- Sent from Reviewable.io -->

jdm commented:
@bors-servo: r+

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11724#-:-KL7P_20PFUQKHkfXP2q:-317921516)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 46c14ac has been approved by `jdm`

<!-- @bors-servo r=jdm 46c14aced2f761c39f0f602962c660e362d98416 -->

bors-servo commented:
:hourglass: Testing commit 46c14ac with merge 54727ed...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/1680)

ConnorGBrewster commented:
@bors-servo retry
 * #11859

bors-servo commented:
:hourglass: Testing commit 46c14ac with merge e43e1eb...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1796)

ConnorGBrewster commented:
@bors-servo retry
 * #11859
 * Infra

bors-servo commented:
:hourglass: Testing commit 46c14ac with merge 8068076...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/1571)

ConnorGBrewster commented:
@bors-servo retry
 * #11859

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/1683), [arm64](http://build.servo.org/builders/arm64/builds/1672), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1707), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1749), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1735), [windows](http://build.servo.org/builders/windows/builds/1408) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/1571), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1804), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1798)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1804)

ConnorGBrewster commented:
@bors-servo retry
 * #11703

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/1683), [arm64](http://build.servo.org/builders/arm64/builds/1672), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1707), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1749), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1735), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1798), [windows](http://build.servo.org/builders/windows/builds/1408) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/1572), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1804)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1805)

ConnorGBrewster commented:
@bors-servo retry
 * #11859 sigh...

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/1572), [arm32](http://build.servo.org/builders/arm32/builds/1683), [arm64](http://build.servo.org/builders/arm64/builds/1672), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1707), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1749), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1735), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1798), [windows](http://build.servo.org/builders/windows/builds/1408) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/1805)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1572), [arm32](http://build.servo.org/builders/arm32/builds/1683), [arm64](http://build.servo.org/builders/arm64/builds/1672), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1707), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1807), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1749), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1735), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1798), [windows](http://build.servo.org/builders/windows/builds/1408)
